### PR TITLE
PARQUET-1338: Fix PrimitiveType.equals throw NPE

### PR DESCRIPTION
**Pull request by wangyum**
Fix `PrimitiveType.equals` throw NPE.

### CODE REVIEW

File: parquet-column/src/main/java/org/apache/parquet/schema/PrimitiveType.java [hunk 1]
```diff
@@ -611,7 +611,7 @@ protected boolean typeEquals(Type other) {
    */
   @Override
   protected boolean equals(Type other) {
-    if (!other.isPrimitive()) {
+    if (other == null || !other.isPrimitive()) {
```
Comment (nandorKollar) [inline parquet-column/src/main/java/org/apache/parquet/schema/PrimitiveType.java]: How come the other is null? Adding a null check sounds reasonable for me, though I wonder in what situation does one call equals with a null parameter.

File: parquet-column/src/main/java/org/apache/parquet/schema/Type.java [hunk 2]
```diff
@@ -283,7 +283,7 @@ protected boolean equals(Type other) {
 
   @Override
   public boolean equals(Object other) {
-    if (!(other instanceof Type) || other == null) {
```
Comment (nandorKollar) [inline parquet-column/src/main/java/org/apache/parquet/schema/Type.java]: Why is this required? The test you wrote fails when I don't apply the change on PrimitiveType, but doesn't fail if I don't apply this.

Comment (nandorKollar) [reply parquet-column/src/main/java/org/apache/parquet/schema/Type.java]: Actually I think null check is not needed here at all: https://docs.oracle.com/javase/specs/jls/se10/html/jls-15.html#jls-15.20.2

**Comment by wangyum** [comment]
It seems that my usage is wrong.

**Comment by rdblue** [comment]
@wangyum, the PrimitiveType's equals method still seems wrong. I think the Type implementation that uses instanceof is fine (because null instanceof Anything is always false) but PrimitiveType is definitely wrong.

**Comment by wangyum** [comment]
Thanks @rdblue. I reopened this.

**Comment by nandorKollar** [comment]
It seems to me that this equals method is called from the public equals method of Type, where the null check is already present. Since null check is already done before (the 'real' equals), what's the advantage of doing it here again?

File: parquet-column/src/main/java/org/apache/parquet/schema/PrimitiveType.java (hunk 1 above)
Comment (rdblue) [reply parquet-column/src/main/java/org/apache/parquet/schema/PrimitiveType.java]: It's part of the method contract: https://docs.oracle.com/javase/7/docs/api/java/lang/Object.html#equals(java.lang.Object)

This basically eliminates the need to add two checks when you want to check equality and the other object may be null. There are lots of cases where equals is used and the code doesn't need to handle null checks in addition.

File: parquet-column/src/test/java/org/apache/parquet/schema/TestTypeBuilders.java [hunk 3]
```diff
@@ -31,6 +31,7 @@
 import static org.apache.parquet.schema.Type.Repetition.*;
 
 public class TestTypeBuilders {
+
```
Comment (rdblue) [inline parquet-column/src/test/java/org/apache/parquet/schema/TestTypeBuilders.java]: Can you remove this? It is a non-functional change.

**Comment by rdblue** [comment]
@nandorKollar, any `equals` implementation should fulfill the documented contract, regardless of how it is used at the moment. Code changes over time and you don't want NPEs thrown because you didn't foresee how something would be used.

**Pull request closed without merging**